content/learning-paths/servers-and-cloud-computing/cca-container/hello-realm.md

```diff
@@ -0,0 +1,132 @@
+
+Install [GCC](/install-guides/gcc/) on your machine. Dependent on the architecture of your machine, you will need to install either the native compiler or cross-compiler.
+
+## Overview
+
+To run an application inside a Realm, you will package it as the "init" process for the guest linux kernel. 
+
+Linux kernels contain a gzipped cpio format archive. When the kernel boots up, this archive is extracted into the root filesystem. The kernel then checks whether the root filesystem contains a file named "init" and tries to run it. At this point the kernel is done booting, and the "init" application executes the system the rest of the way. 
+
+In this section, you will package a simple hello world application into the initramfs for the guest linux kernel. 
+
+## Create a simple initramfs for the guest linux kernel
+
+Create a directory in which you will build the simple initramfs:
+
+```console
+mkdir ~/FM-share
+cd ~/FM-share
+```
+Using a file editor of your choice, create a file named `hello.c` in this directory. Copy the contents below into the file:
+
+```console
+#include <stdio.h>
+#include <unistd.h>
+
+int main(int argc, char *argv[])
+{
+  printf("Hello from the Realm!\n");
+  sleep(99);
+}
+```
+
+Now compile the application as the init executable and package it into the `initramfs`:
+
+```console
+aarch64-linux-gnu-gcc -static hello.c -o init
+echo init | cpio -o -H newc | gzip > test.cpio.gz
+```
+`test.cpio.gz` is the compressed archive you will use as the initramfs for the guest linux kernel.
+
+## Run the application inside the Realm 
+
+Start the docker container with the `FM-share` directory mounted:
+
+```console
+docker run -v ~/FM-share:/home/ubuntu/FM-share -it armswdev/aemfvp-cca-image /bin/bash
+```
+Run the Arm CCA reference stack pre-built binaries on the FVP:
+
+```console
+./run-cca-fvp.sh
+```
+After the host linux kernel has booted up, you will prompted to login. Use `root` for both the login and password:
+
+```output
+[    3.879183] VFS: Mounted root (ext4 filesystem) readonly on device 254:0.
+[    3.879643] devtmpfs: mounted
+[    3.897918] Freeing unused kernel memory: 9024K
+[    3.901685] Run /sbin/init as init process
+[    3.965637] EXT4-fs (vda): re-mounted 34fdcd00-4506-42fc-8af8-a35ca9cf965d r/w. Quota mode: none.
+Starting syslogd: OK
+Starting klogd: OK
+Running sysctl: OK
+Saving random seed: OK
+Starting network: ip: RTNETLINK answers: File exists
+udhcpc: started, v1.31.1
+udhcpc: sending discover
+udhcpc: sending select for 172.20.51.1
+udhcpc: lease of 172.20.51.1 obtained, lease time 86400
+deleting routers
+adding dns 172.20.51.254
+FAIL
+Starting dropbear sshd: OK
+
+Welcome to Buildroot
+buildroot login: root
+Password:
+#
+```
+With the host linux kernel running, you can now enable sharing of files between your host machine and the running simulation. Mount a `9p` drive on the host linux kernel using `/etc/fstab`:
+
```

Comment:
This step is a bit fiddly - I wonder if it could be automated?

Reply:
I did look into this briefly, couldn't automate it with the few experiments I tried. Hence the manual step to mount. When I create the next rev of the container, I'll check if there's some file in the buildroot I can add this too where it won't get overwritten as part of the build.